src/transformers/models/bloom/modeling_bloom.py

```diff
@@ -760,6 +770,12 @@ def __init__(self, config):
         # Initialize weights and apply final processing
         self.post_init()
 
+    def tie_weights(self):
```

Comment:
Don't think that this is better than adding an if statement to the forward. To me upcasting the lm_head is similar/identical to upcasting the attention softmax to make it more stable which happens in the forward. 

From a user's perspective, I would expect the weights to be downloaded and loaded into the model **exactly** how they are on the Hub and not changed on the fly depending on a config - think this is quite confusing. 

Then during the forward the operation, can't we just force the outputs of the model to be in float32 depending on the config like we do in GPTJ: https://github.com/huggingface/transformers/blob/1e380c7dcb5af5b6820f363548c44ac09feac352/src/transformers/models/gptj/modeling_gptj.py#L844 
I'm also fine with upcasting the weights in the forward then doing the `nn.Linear(...)` operation and then downcasting them again, so that weights dtype before `forward()` == weights dtype after `forward()`. It would be nice that neither loading the weights into the model nor the forward pass actually changes the weights dtype.

Reply:
Also just don't like overwriting functions like `tie_weights` from `PreTrainedModel` this makes the code hard to read (e.g. just the loading of the model weights now switches back multiple times between files)

@sgugger @stas00 what do you think here?

Reply:
Forcing the logits to be in float32 instead of the whole lm head would be way better in terms of memory management (if it works). That's a change we could have without opt-in.

Reply:
> From a user's perspective, I would expect the weights to be downloaded and loaded into the model exactly how they are on the Hub and not changed on the fly depending on a config - think this is quite confusing.

But that's not true, if you don't specify `torch_dtype` it gets loaded as fp32, even when the weights are in another precision.

> Then during the forward the operation, can't we just force the outputs of the model to be in float32 depending on the config like we do in GPTJ:

This shouldn't work as you run `nn.Linear` in fp16/bf16 and it collapses at that point and then cast fo fp32, but it's already collapsed. I don't think that `nn.Linear` allows to output tensors in another precision than the input.